protocol/src/main/java/org/sdo/pri/OnDieSignatureValidator.java

```diff
@@ -0,0 +1,185 @@
+      return false;
+    }
+
+    // Now validate the signature
+    // Public key comes from the cert chain in the voucher.
+    return validateWithoutRevocations(signedData, signature, certificateList.get(0).getPublicKey());
+  }
+
+
+  private static byte[] convertSignature(byte[] signature, byte[] taskInfo) throws Exception {
+    // 0x30 b1 0x02 b2 (vr) 0x02 b3 (vs)
+    byte[] rvalue = Arrays.copyOfRange(signature, taskInfo.length, taskInfo.length + 48);
+    byte[] svalue = Arrays.copyOfRange(signature, taskInfo.length + 48, taskInfo.length + 96);
+
+    boolean appendZeroToR = false;
+    boolean appendZeroToS = false;
+    if ((rvalue[0] & 0x80) != 0) {
+      appendZeroToR = true;
+    }
+    if ((svalue[0] & 0x80) != 0) {
+      appendZeroToS = true;
+    }
+
+    ByteArrayOutputStream adjSignature = new ByteArrayOutputStream();
+    adjSignature.write(0x30);
+    // total length of remaining bytes
+    adjSignature.write(4 + (appendZeroToR ? 49 : 48) + (appendZeroToS ? 49 : 48));
+    adjSignature.write(0x02);
+    // R value
+    if (appendZeroToR) {
+      adjSignature.write(49);
+      adjSignature.write(0x00);
+      adjSignature.write(rvalue);
+    } else {
+      adjSignature.write(48);
+      adjSignature.write(rvalue);
+    }
+    adjSignature.write(0x02);
+    // S value
+    if (appendZeroToS) {
+      adjSignature.write(49);
+      adjSignature.write(0x00);
+      adjSignature.write(svalue);
+    } else {
+      adjSignature.write(48);
+      adjSignature.write(svalue);
+    }
+    return adjSignature.toByteArray();
+  }
+
+
+  private static boolean checkRevocations(
+          List<Certificate> certificateList,
+          OnDieCache onDieCache) {
+    // Check revocations first.
+    try {
+      CertificateFactory certificateFactory = CertificateFactory.getInstance("X509");
+      for (int i = 0; i < certificateList.size(); i++) {
+        X509Certificate cert = (X509Certificate) certificateList.get(i);
+        X509CertificateHolder certHolder = new X509CertificateHolder(cert.getEncoded());
+        CRLDistPoint cdp = CRLDistPoint.fromExtensions(certHolder.getExtensions());
+        if (cdp != null) {
+          DistributionPoint[] distPoints = cdp.getDistributionPoints();
+          for (DistributionPoint dp : distPoints) {
+            GeneralName[] generalNames =
+                    GeneralNames.getInstance(dp.getDistributionPoint().getName()).getNames();
+            for (int j = 0; j < generalNames.length; j++) {
+              byte[] crlBytes = onDieCache.getCrl(generalNames[j].getName().toString());
+              if (crlBytes == null) {
+                LoggerFactory.getLogger(OnDieSignatureValidator.class).warn(
+                        "CRL not found in cache for: " + generalNames[j].getName().toString());
+                return false;
+              } else {
+                CRL crl = certificateFactory.generateCRL(new ByteArrayInputStream(crlBytes));
+                if (crl.isRevoked(cert)) {
```

Comment:
what if 'isRevoked' returns one of the odd cases like 'unknown'?

Reply:
this is always true/false. Revocations are always checked against local crl files. There is no OCSP as well.